### PR TITLE
[axonious-237] Initial Commit

### PR DESCRIPTION
**Pull request by Phillips-Chris**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
N/A

## Description
Initial commit of Axonius integration.

## Screenshots
N/A

## Minimum version of Demisto
- [ ] 5.0.0
- [X] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [ ] Tests
- [ ] Documentation 

## Demisto Partner?
- [X] axonious-237

### CODE REVIEW

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    max_rows: int = get_int_arg(key="max_results", required=False, default=MAX_ROWS)
+
+    assets = api_obj.get_by_saved_query(name=name, max_rows=max_rows)
+    results = [parse_asset(asset=asset) for asset in assets]
+
+    return CommandResults(
+        outputs_prefix=f"Axonius.{api_obj.__class__.__name__}",
+        outputs_key_field="internal_axon_id",
+        outputs=results,
+    )
```
Comment (guyfreund) [inline Packs/Axonius/Integrations/Axonius/axonius.py:89]: You should provide a human readable outputs in `markdown` format, you should also make sure there's a human readable output when no assets are present. You can use `tableToMarkdown` from `CommonServerPython` script, See [tableToMarkdown Docs](https://xsoar.pan.dev/docs/integrations/code-conventions#tabletomarkdown) for more information.
Remember to differentiate users hr from devices hr.
```suggestion
        readable_output=hr
    )
```

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+def get_by_sq(
+    api_obj: axonius_api_client.api.assets.asset_mixin.AssetMixin,
+) -> CommandResults:
+    """Get assets with their defined fields returned by a saved query."""
+    args: dict = demisto.args()
+    name: str = args["saved_query_name"]
+    max_rows: int = get_int_arg(key="max_results", required=False, default=MAX_ROWS)
+
+    assets = api_obj.get_by_saved_query(name=name, max_rows=max_rows)
+    results = [parse_asset(asset=asset) for asset in assets]
```
Comment (guyfreund) [inline Packs/Axonius/Integrations/Axonius/axonius.py:83]: You should handle the case where no assets are present.

File: Packs/Axonius/Integrations/Axonius/axonius.yml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  required: true
+  type: 0
+- name: ax_secret
+  display: Axonius API Secret
+  required: true
+  type: 4
+- name: insecure
+  display: Trust any certificate (not secure)
+  required: false
+  type: 8
```
Comment (guyfreund) [inline Packs/Axonius/Integrations/Axonius/axonius.yml:26]: I've seen you supported in proxy in the main function, if so please add a parameter for the integration to make it work.
```suggestion
- name: insecure
  display: Trust any certificate (not secure)
  required: false
  type: 8
- display: Use system proxy settings
  name: proxy
  required: false
  type: 8
```

**Review by guyfreund — CHANGES_REQUESTED** [review]
Overall looking good, several changes to be done here and some recommendations.

#### Changes:
- You should add an integration documentation file (`Packs/Axonius/Integrations/Axonius/README.md`), see [Integration Docs](https://xsoar.pan.dev/docs/integrations/integration-docs) for more information. When you write the docs, don't forget to add the `XSOAR` version the integration was tested with.
- You should change the name of the integration `yml` file: `axonius.yml` => `Axonius.yml`
- We recommend adding content to the Pack `README.md` file (`Packs/Axonius/README.md`), see [Pack Docs](https://xsoar.pan.dev/docs/integrations/pack-docs) for more information.
- We recommend adding content to the Integration description file (`Packs/Axonius/Integrations/Axonius/Axonius_description.md`) to contain details on how to configure an instance of the integration.
- Its better practice to pass the client object instead of API objects, but that's not enforced by us.

File: Packs/Axonius/Integrations/Axonius/Axonius_description.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+## Hello World
+- This text is markdown
+- This section explains how to configure the instance of HelloWorld in Cortex XSOAR.
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/Axonius_description.md:3]: Updated with relevant information.

File: Packs/Axonius/Integrations/Axonius/axonius.yml [hunk 5]
```diff
@@ -0,0 +1,102 @@
+name: Axonius
+display: Axonius
+description: This is the Axonius integration
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.yml:3]: Updated description accordingly.

File: Packs/Axonius/Integrations/Axonius/axonius.yml [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  subtype: python3
+  dockerimage: demisto/axonius:1.0.0.10828
+  runonce: false
+  commands:
+  - name: axonius-get-devices-by-savedquery
+    description: Gather device info by saved query
+    arguments:
+    - name: saved_query_name
+      required: true
+      description: The name of the saved query
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.yml:39]: Updated accordingly.

File: Packs/Axonius/Integrations/Axonius/axonius.yml [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      description: The OS type (Windows, Linux, iOS,...)
+    - contextPath: Axonius.Devices.labels
+      type: String
+      description: Tags assigned to the asset
+  - name: axonius-get-users-by-savedquery
+    description: Gather user info by saved query
+    arguments:
+    - name: saved_query_name
+      required: true
+      description: The name of the saved query
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.yml:76]: Updated accordingly.

File: Packs/Axonius/Integrations/Axonius/axonius.yml [hunk 8]
```diff
@@ -0,0 +1,102 @@
+name: Axonius
+display: Axonius
+description: This is the Axonius integration
+category: Data Enrichment & Threat Intelligence
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.yml:4]: @guyfreund We didn't know what to fill in for this value, we only have 5.5 locally so I just put what we were using. Should we not specify a version here?

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
+"""Demisto Integration for Axonius."""
+import traceback
+from typing import Any, List, Optional
+
+import axonius_api_client
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py:8]: @guyfreund We noticed that the 'demisto-sdk upload' command strips those lines out of the code no matter where they are. Our linter (isort) forces the correct order of imports which is not having those at the top of the script.

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    if "last_seen" in key:
+        try:
+            return axonius_api_client.tools.dt_parse(value).isoformat()
+        except Exception:
+            return value
+    return value
+
+
+def parse_key(key: str) -> str:
+    """Pass."""
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py]: Updated accordingly.

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        key = axonius_api_client.tools.strip_left(obj=key, fix="specific_data.data.")
+        key = f"aggregated_{key}"
+    if key.startswith("adapters_data."):
+        key = axonius_api_client.tools.strip_left(obj=key, fix="adapters_data.")
+    key = key.replace(".", "_")
+    return key
+
+
+def parse_asset(asset: dict) -> dict:
+    """Pass."""
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py]: Updated accordingly.

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    try:
+        return int(value)
+    except Exception:
+        raise ValueError(
+            f"Supplied value {value!r} for argument {key!r} is not an integer."
+        )
+
+
+def test_module(client: axonius_api_client.connect.Connect) -> str:
+    """Pass."""
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py]: Updated accordingly.

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    client.start()
+    return "ok"
+
+
+def parse_kv(key: str, value: Any) -> Any:
+    """Pass."""
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py]: Updated accordingly.

File: Packs/Axonius/Integrations/Axonius/axonius.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (guyfreund) [reply Packs/Axonius/Integrations/Axonius/axonius.yml:4]: @Phillips-Chris since your pack doesn't use 6.0 only features I'd put `5.0.0`.

**Review by guyfreund — CHANGES_REQUESTED** [review]
Please submit all other requested changes

File: Packs/Axonius/pack_metadata.json [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+    "url": "https://docs.axonius.com",
+    "email": "support@axonius.com",
+    "created": "2020-07-30T18:52:36Z",
+    "categories": [
+        "Data Enrichment & Threat Intelligence"
+    ],
+    "tags": [],
+    "useCases": [],
+    "keywords": [],
+    "githubUser": "Phillips-Chris"
```
Comment (guyfreund) [inline Packs/Axonius/pack_metadata.json:16]:
```suggestion
    "githubUser": ["Phillips-Chris"]
```

**Review by guyfreund — CHANGES_REQUESTED** [review]

**Comment by guyfreund** [comment]
@Phillips-Chris any updates regarding the needed changes?

File: Packs/Axonius/Integrations/Axonius/axonius.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.yml:4]: Updated with correct version info

File: Packs/Axonius/Integrations/Axonius/axonius.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        return int(value)
+    except Exception:
+        raise ValueError(
+            f"Supplied value {value!r} for argument {key!r} is not an integer."
+        )
+
+
+def test_module(client: axonius_api_client.connect.Connect) -> str:
+    """Pass."""
+    client.start()
```
[the remark this answers is not in the document]
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py:40]: @guyfreund Yes it does.

File: Packs/Axonius/pack_metadata.json (hunk 14 above)
Comment (Phillips-Chris) [reply Packs/Axonius/pack_metadata.json]: Updated this to the proper form

File: Packs/Axonius/Integrations/Axonius/axonius.py (hunk 1 above)
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py:89]: @guyfreund @fvigo said that our data would be automatically formatted. We went through a few examples together to make sure it formatted correctly. During that call, we collectively decided not to manually setup the readable output.

File: Packs/Axonius/Integrations/Axonius/axonius.py (hunk 2 above)
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/axonius.py:80]: Added logic check for no assets returned.

File: Packs/Axonius/Integrations/Axonius/Axonius.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+        for k, v in asset.items()
+        if k not in SKIPS
+    }
+
+
+def check_asset_count(api_obj: Union[Users, Devices], assets: List[dict], src: str):
+    """Raise exception for no assets found."""
+    if not assets:
+        api_name = api_obj.__class__.__name__
+        raise Exception(f"No {api_name} assets returned from {src}.")
```
Comment (guyfreund) [inline Packs/Axonius/Integrations/Axonius/Axonius.py:93]: don't raise an Excpetion here, just return the expected human readable to the command and build a CommandResults object with no outputs and the human readable returned.

File: Packs/Axonius/Integrations/Axonius/Axonius.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+    try:
+        return int(value)
+    except Exception:
+        raise ValueError(
+            f"Supplied value {value!r} for argument {key!r} is not an integer."
+        )
+
+
+def get_csv_arg(
```
Comment (guyfreund) [inline Packs/Axonius/Integrations/Axonius/Axonius.py:39]: You can use `argToList` function from `CommonServerPython` imported at the beginning of this py file.

**Review by guyfreund — CHANGES_REQUESTED** [review]
Looking good, to be done here:
- Enhance the integration's documentation (`Packs/Axonius/Integrations/Axonius/README.md`) with the added commands.
- Update the commands file (`Packs/Axonius/Integrations/Axonius/command_examples.txt`) with the added commands.
- Why have you added so much `noqa` disable statements?

File: Packs/Axonius/Integrations/Axonius/Axonius.py (hunk 16 above)
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/Axonius.py:93]: @guyfreund When working with @fvigo we were told to raise an exception when no assets were found. Wouldn't the human readable output for this just be nothing or an empty list? We can just return an empty list if you would rather we didn't raise an exception.

File: Packs/Axonius/Integrations/Axonius/Axonius.py (hunk 17 above)
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/Axonius.py:39]: @guyfreund Updated to use argToList in our function while still checking for empty list items. The argToList function has a short coming in that it does not remove "empty" strings produced by input like "a,b,,c,d". We coded around that.

File: Packs/Axonius/Integrations/Axonius/Axonius.py (hunk 16 above)
Comment (guyfreund) [reply Packs/Axonius/Integrations/Axonius/Axonius.py:93]: @Phillips-Chris no need to raise an exception. the human readable should be `No assets were found.`

File: Packs/Axonius/Integrations/Axonius/Axonius.py (hunk 17 above)
Comment (guyfreund) [reply Packs/Axonius/Integrations/Axonius/Axonius.py:39]: OK, whatever is good for you

File: Packs/Axonius/Integrations/Axonius/Axonius.py (hunk 16 above)
Comment (Phillips-Chris) [reply Packs/Axonius/Integrations/Axonius/Axonius.py:93]: @guyfreund Got it, updated the code appropriately.

**Review by guyfreund — APPROVED** [review]

**Comment by guyfreund** [comment]
@Phillips-Chris You have some more sensitive data in the command examples file, not sure if you want to delete it or not.

**Comment by Phillips-Chris** [comment]
> @Phillips-Chris You have some more sensitive data in the command examples file, not sure if you want to delete it or not.
> This PR can be merged from our end.

It is example data that is safe to be in the sample data. Thanks!

**Pull request merged**